Adds ramp acceleration to FREEZE_FEATURE

Switchleg1 opened the pull request:
Simply removes the FREEZE_FEATURE instant halt code and adds a linear ramp of deceleration when pin is triggered and acceleration back to full speed on deactivation.  Acceleration values are based on the current planner blocks and final jerk is set within the configuration file.

This is achieved by intercepting the calculated step_rate during Step::block_phase_isr() on each planner block stages (init/accel/decel/cruise) by tracking the time accumulated since triggering the freeze pin.


For example here is the deceleration implementation:

#if ENABLED(FREEZE_FEATURE)
  if(frozen_time) check_frozen_time(step_rate);
#endif

// step_rate to timer interval and steps per stepper isr
interval = calc_multistep_timer_interval(step_rate << oversampling_factor);
deceleration_time += interval;

#if ENABLED(FREEZE_FEATURE)
  check_frozen_pin(1, interval);
#endif

-->

### Requirements

Should work on all boards, has been tested on mega2560.

### Benefits

Prevents skipped steps when trying to pause a machine.

### Configurations

Configuration_Adv.h
-----------------------
#define FREEZE_FEATURE
#if ENABLED(FREEZE_FEATURE)
  #define FREEZE_PIN      5     // Override the default (KILL) pin here
  #define FREEZE_JERK     2     // Completely halt when motion has decelerated below this value 
  #define FREEZE_STATE    LOW   // State of pin indicating freeze
#endif

adds FREEZE_JERK as in the example above

Switchleg1 committed:
freeze_feature

This allows the freeze_feature to use acceleration when triggering freeze_pin

Switchleg1 committed:
update

Switchleg1 committed:
corrects missing 'acceleration_rate'

Switchleg1 committed:
Adds acceleration_rate in planner.h when S_CURVE acceleration is being used.

Switchleg1 commented:
I apologize for the multiple updates.  Copying over from another branch it appears I missed a few lines.

Switchleg1 committed:
Optimizations

Prevents calculating unnecessary linear acceleration rate during path planning when using s_curve_acceleration.  General cleanup and commented out feature by default in configuration.

thisiskeithb commented:
This wording will also need to be adjusted (or options/defaults changed) since it will no longer be "instant": https://github.com/MarlinFirmware/Marlin/blob/4eb8a871930ef152e4fa73fc1782e2ed22348c85/Marlin/Configuration_adv.h#L4292

Switchleg1 committed:
Adds option to turn off laser

User has the option of turning the laser off when motion is halted regardless of laser mode.  Once motion starts previous laser power is restored.

Also makes changes to feature description in config_adv.h

Switchleg1 committed:
optimizations and bug fixes

If freeze is triggered when machine is stationary it immediately sets solid flag.

thinkyhead committed:
Format

thinkyhead committed:
general cleanup, test with s-curve

thinkyhead committed:
add'l format

thinkyhead commented on Marlin/src/MarlinCore.cpp:
```diff
@@ -500,7 +500,7 @@ inline void manage_inactivity(const bool no_stepper_sleep=false) {
 
   #if ENABLED(FREEZE_FEATURE)
-    stepper.frozen = READ(FREEZE_PIN) == FREEZE_STATE;
+    stepper.set_frozen_triggered(READ(FREEZE_PIN) == FREEZE_STATE);
```
Note that without some debounce this state will oscillate, and will be reset if the pin is released. It will be simpler to set the frozen state with a single trigger of the pin and then not reset it to 0 when the pin is un-triggered, but if that is too sensitive then some debounce logic can help. It depends on the use case for freeze and the kind of switch that it's attached to.

Switchleg1 replied:
Yes though bouncing only affects whether the machine is accelerating or decelerating for the duration of the bounce and since it does not violate the machine abilities do we care?  I have been using similar code on my CNC router using a standard rocker switch on ramps/AVR for the past month and a half and it's served me well without any noticable stuttering.

Switchleg1 replied:
![image (1)](https://github.com/user-attachments/assets/6e0b874f-8aea-4523-af30-9cc346a9cd16)

Example of how a bounce would look.

Switchleg1 replied:
We also want to be able to restart the machine after a freeze condition.  Some scenarios include material build up in the tool,  also incorrect gcode where we would like to clear the current buffer (and current move) and restart without having to redial the machine back in as in the case of using the kill method.  I had implemented this in repetier years ago and plan to complete this function in Marlin when I have a chance.

thinkyhead committed:
refine config

thinkyhead committed:
fix test

Switchleg1 committed:
drill cycles

Switchleg1 committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into bugfix-2.1.x

Switchleg1 committed:
add mention to M115 and correct linking based on feature
